### PR TITLE
Tutorial: toy arbitration logic

### PR DESCRIPTION
**Pull request by anshumanmohan**
This PR adds a handwritten example of arbitration logic. 

It will be nice to have this at hand during the tutorial, since arbitration is a key step towards lifting the following restriction of MrXL's:
> If repeated `map`/`reduce` operations are performed on the same array, each of those operations must have the same parallelism factor.

The way you do it is: 
1. See which parallelism factors are applied to that array (say, 2 and 3).
2. Actually bank it per the LCM of those factors (so, 6).
3.  Develop arbitration logic to maintain the fiction that the array is banked this way or that (so, as 2 or 3).

The handwritten example `arb_6_as_3_or_2.futil` shows how to take six memories of size 4 each and then pretend, simultaneously, that:
- They are actually _two_ memories of size _12_ each.
- They are actually _three_ memories of size _8_ each.

The development is tellingly in the `calyx-py` directory because I'd like to eventually generate this via the `builder` library. An attempt to generate it with the builder is also checked in; see `arb.py`. Heads up that `arb.py` is trying to inch towards `arb_hw.futil`, which is a simpler handwritten version, and not towards `arb_6_as_3_or_2.futil`, the full handwritten version.

There are comments throughout. In `arb_6_as_3_or_2.futil` the comments are mostly pedagogical, in case a tutorial user does end up reading this stuff directly. In `arb.py` my comments explain my roadblocks.

### CODE REVIEW

File: calyx-py/test/arb_hw_6.futil [hunk 1]
```diff
@@ -0,0 +1,373 @@
+import "primitives/core.futil";
+
+component wrap2(i: 32, j: 32) -> () {
```
Comment (sampsyo) [inline calyx-py/test/arb_hw_6.futil:3]: You might consider calling this `arb_6_to_12` or similar to convey its purpose a bit.

**Review by sampsyo — COMMENTED** [review]
Really nice work on this!! I think this neatly pulls together many concepts in Calyx, especially `ref` cells, in a way that is pedagogically useful. It also obviously makes the case for generating the code instead of writing it by hand, because each of the differently sized arbiters looks pretty much the same. Neato!

File: calyx-py/test/arb.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    mem2 = wrap.mem_d1("mem2", 32, 4, 32, is_ref=True)
+    ans = wrap.mem_d1("ans", 32, 1, 32, is_ref=True)
+
+    add_eq(wrap, "i", 0, "eq0", "i_eq_0")
+    add_eq(wrap, "i", 1, "eq1", "i_eq_1")
+    add_emit_from_mem(wrap, mem1, ans, "1")
+    add_emit_from_mem(wrap, mem2, ans, "2")
+
+    # AM, quality of life:
+    # I'd like to generate these `if` statements with the builder.
```
Comment (rachitnigam) [inline calyx-py/test/arb.py:60]: Useful to point this out to @EclecticGriffin as a potential `builder` improvement

File: calyx-py/test/arb.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    # Maybe I'm missing something, but I think the builder library
+    # is only targeting a subset of the `invoke` functionality.
+    #   class Invoke(Control):
+    #     id: CompVar
+    #     in_connects: List[Tuple[str, Port]]
+    #     out_connects: List[Tuple[str, Port]]
+    #     ref_cells: List[Tuple[str, CompVar]] = field(default_factory=list)
+    #     comb_group: Optional[CompVar] = None
+    #     attributes: List[Tuple[str, int]] = field(default_factory=list)
+    # As I see it, only id, in_connects, and out_connects are supported.
```
Comment (rachitnigam) [inline calyx-py/test/arb.py:122]: @anshumanmohan can you explain this comment? Are the other fields missing from the dataclass definition or are they unused?

File: calyx-py/test/arb.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    _ = main.mem_d1("A", 32, 4, 32, is_external=True)
+    _ = main.mem_d1("B", 32, 4, 32, is_external=True)
+    _ = main.mem_d1("out", 32, 1, 32, is_external=True)
+
+    # AM, point of failure (but a stupid hack has worked):
+    # I'd like to add the following to the `cells` section:
+    # together = wrap();
+
+    # The following is clearly a hilarious hack:
+    _ = main.cell("together", CompVar("wrap()"))
```
Comment (rachitnigam) [inline calyx-py/test/arb.py:103]: Another thing for @EclecticGriffin potentially!

**Review by rachitnigam — APPROVED** [review]
Really cool work @anshumanmohan! I think this implementation is really clean and also a good example of why `ref` cells are useful (something that we hadn't quite pinned down up till now). What do you think about:

1. Adding the new examples to `runt` testing? I think it would be pretty straightforward
2. Point to this example in the [`ref` cell docs](https://github.com/cucapra/calyx/blob/master/docs/lang/memories-by-reference.md)?

I would also briefly wait and get an acknowledgement from @EclecticGriffin that they've seen the comments you've left in the file because they are useful things to be added to `calyx-py`.

**Comment by rachitnigam** [comment]
Also @paili0628 @calebmkim, I think we should talk about `ref` cells in the Calyx 2.0 paper and maybe use the example of arbiters from here as a reason why `ref` syntax is useful

File: calyx-py/test/arb.py (hunk 3 above)
Comment (anshumanmohan) [reply calyx-py/test/arb.py:122]: Unused! If you look at the relevant `builder` code:
https://github.com/cucapra/calyx/blob/2613b848efa3b6344edd2cd9ed3c722c889fb532/calyx-py/calyx/builder.py#L221-L239

You can see it essentially returning `ast.Invoke(id, in_connects, out_connects)`. It's leaving the other `ast.Invoke` fields blank, and then the `ast.Invoke` `doc()` method is filling in default no-ops.

File: calyx-py/test/arb.py (hunk 2 above)
Comment (sampsyo) [reply calyx-py/test/arb.py:60]: FWIW, it should be possible to do this with `cb.if_` when using the values returned from building the stuff. That is, this should hypothetically work:

```py
cb.if_(eq0.out, i_eq_0, emit_from_mem1)
```

But that will require making, for example, the `add_eq` function return a value (`eq_cell`) and putting that result into the variable `eq0`. But that should simplify things a bit vs. using the string names everywhere.

File: calyx-py/test/arb.py (hunk 4 above)
Comment (sampsyo) [reply calyx-py/test/arb.py:103]: Some relevant advice here: https://github.com/cucapra/calyx/pull/1487#discussion_r1200814552

File: calyx-py/test/arb.py (hunk 3 above)
Comment (sampsyo) [reply calyx-py/test/arb.py:122]: Yes, this is right—I just never hooked up the ref-cell stuff in this simple wrapper. We should probably do that somehow!

**Review by sampsyo — COMMENTED** [review]
Even more hints!

**Comment by anshumanmohan** [comment]
I have now developed `arb6.py`, which is inching towards the big kahuna: `arb_6_as_3_or_2.py`. I have run into two issues:
1. I don't know how to use the builder to generate `if a with b then c else d`-style code.
2. I don't think I'm using @EclecticGriffin 's changes to add `invoke` correctly. It's not checked in, but try the following in the `control` section of `wrap2`:

        main.control += [
          cb.invoke(together2, in_i=cb.const(32, 1), in_j=cb.const(32, 11), ref_mem1=A),
          cb.invoke(together3, in_i=cb.const(32, 2), in_j=cb.const(32, 7)),
        ]
I get `TypeError: 'ExprBuilder' object is not callable`, which makes sense: `A` is a `CellBuilder` and `cb.invoke` is going to immediately run `ExprBuilder.unwrap(v)` on it. There's something more to it which I'm missing.

**Comment by anshumanmohan** [comment]
@rachitnigam since the Python-generated code is so close, I'm going to wait a day for the last of it to land (see the comment just above) and then I'll add that to `runt` testing!

**Pull request merged**